File: src/test/java/com/sqat/pages/SearchResultsPage.java
package com.sqat.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.TimeoutException;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.testng.SkipException;

import java.time.Duration;
import java.util.List;

public class SearchResultsPage extends BasePage {
    
    // Locators
    private static final String FIRST_PRICE_CSS = "div[data-test-id='price']";
    // Buy button for Wingie partner only
    private static final By WINGIE_BUY_BUTTON = By.xpath(
        "//div[@data-test-id='text' and contains(., 'Wingie')]" +
        "/ancestor::div[contains(@data-test-id,'proposal') or contains(@class,'proposal')]" +
        "//button"
    );

    public SearchResultsPage(WebDriver driver) {
        super(driver);
    }

    /**
     * Open the search results URL
     */
    public void openSearchUrl(String url) {
        driver.get(url);
        waitForPageLoad();
    }

    /**
     * Verify that the page has loaded by checking title
     */
    public String verifyPageLoaded() {
        waitForPageLoad();
        String title = getPageTitle();
        assert !title.trim().isEmpty() : "Page title is empty";
        return title;
    }

    /**
     * Wait for manual captcha solving
     */
    public void waitForCaptcha(int waitSeconds) {
        try {
            Thread.sleep(waitSeconds * 1000L);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    /**
     * Select the first available ticket by clicking on its price
     */
    public void selectFirstTicket() {
        WebElement firstPrice = wait.until(
            ExpectedConditions.elementToBeClickable(By.cssSelector(FIRST_PRICE_CSS))
        );
        safeClick(firstPrice);
    }

    /**
     * Click the buy button and return the original window handle
     * Only clicks if the offer is from Wingie, otherwise skips the test.
     */
    public String clickBuyButton() {
        WebElement buyButton;
        try {
            buyButton = wait.until(
                ExpectedConditions.elementToBeClickable(WINGIE_BUY_BUTTON)
            );
        } catch (TimeoutException e) {
            // Wingie offer not available – skip the test
            throw new SkipException("Wingie offer not found on Aviasales search results, skipping test", e);
        } catch (Exception e) {
            throw new SkipException("Unable to locate Wingie buy button, skipping test", e);
        }

        String originalWindow = driver.getWindowHandle();
        safeClick(buyButton);
        return originalWindow;
    }

    /**
     * Switch to the partner booking window
     */
    public void switchToPartnerWindow(String originalWindow) {
        WebDriverWait shortWait = new WebDriverWait(driver, Duration.ofSeconds(15));
        shortWait.until(ExpectedConditions.numberOfWindowsToBe(2));
        
        for (String handle : driver.getWindowHandles()) {
            if (!handle.equals(originalWindow)) {
                driver.switchTo().window(handle);
                break;
            }
        }
        
        // Wait for page to fully load
        waitForPageLoad();
        
        // Additional wait for redirect page to complete
        try {
            Thread.sleep(3000); // Wait for redirect
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        
        // Wait for booking form to appear (check for any booking form element)
        WebDriverWait formWait = new WebDriverWait(driver, Duration.ofSeconds(30));
        try {
            // Try to wait for email field or any booking form element
            formWait.until(ExpectedConditions.presenceOfElementLocated(
                By.cssSelector("input[id*='email'], input[id*='Email'], input[id*='contact'], form")
            ));
        } catch (Exception e) {
            // If specific element not found, just wait a bit more
            try {
                Thread.sleep(5000);
            } catch (InterruptedException ie) {
                Thread.currentThread().interrupt();
            }
        }
    }
}

File: pages/search_results_page.py
from selenium.webdriver.common.by import By
from selenium.webdriver.support import expected_conditions as EC
from pages.base_page import BasePage
import time


class SearchResultsPage(BasePage):
    """Page object for Aviasales search results page"""

    # Locators
    FIRST_PRICE_SELECTOR = "div[data-test-id='price']"
    BUY_BUTTON_XPATHS = [
        "//button[contains(.,'Купить')]",
        "//button[contains(.,'Buy')]",
        "(//button[contains(@class,'button')])[1]"
    ]

    def __init__(self, driver, wait_timeout=25):
        super().__init__(driver, wait_timeout)

    def open_search_url(self, url):
        """Open the search results URL"""
        self.driver.get(url)
        self.wait_for_page_load()

    def verify_page_loaded(self):
        """Verify that the page has loaded by checking title"""
        title = self.get_page_title()
        assert title.strip() != "", "Page title is empty"
        return title

    def wait_for_captcha(self, wait_seconds=30):
        """Wait for manual captcha solving"""
        print(f"ACTION REQUIRED: Solve captcha manually if it appears ({wait_seconds} sec)")
        time.sleep(wait_seconds)

    def select_first_ticket(self):
        """Select the first available ticket by clicking on its price"""
        first_price = self.wait.until(
            EC.element_to_be_clickable((By.CSS_SELECTOR, self.FIRST_PRICE_SELECTOR))
        )
        self._safe_click(first_price)
        print("Checkpoint 2: First ticket price clicked")

    def click_buy_button(self):
        """Click the buy button and return the original window handle"""
        buy_button = None
        for xpath in self.BUY_BUTTON_XPATHS:
            try:
                buy_button = self.wait.until(EC.element_to_be_clickable((By.XPATH, xpath)))
                break
            except Exception:
                continue

        if buy_button is None:
            raise Exception("Buy button not found")

        original_window = self.driver.current_window_handle
        self._safe_click(buy_button)
        print("Checkpoint 3: Buy button clicked")
        return original_window

    def switch_to_partner_window(self, original_window, timeout=15):
        """Switch to the partner booking window"""
        WebDriverWait(self.driver, timeout).until(EC.number_of_windows_to_be(2))
        for handle in self.driver.window_handles:
            if handle != original_window:
                self.driver.switch_to.window(handle)
                break
        self.wait_for_page_load()
        print("Checkpoint 4: Partner page opened ->", self.get_page_title())

